pkg/test/externalbinary/provider.go

```diff
@@ -0,0 +1,194 @@
+package externalbinary
+
+import (
+	"fmt"
+	"log"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+	"time"
+
+	imagev1 "github.com/openshift/api/image/v1"
+	"github.com/pkg/errors"
+
+	"github.com/openshift/origin/test/extended/util"
+)
+
+// ExternalBinaryProvider handles extracting external test binaries from a given payload. By
+// default, it uses a cache directory for extracted binaries assuming they'll be reused,
+// especially when developing locally. Set OPENSHIFT_TESTS_DISABLE_CACHE to any non-empty
+// value to use a temporary directory instead that will be removed at end of execution. When
+// using caching, files older than 7 days will be removed.
+type ExternalBinaryProvider struct {
+	oc                   *util.CLI
+	binPath              string
+	tmpDir               string
+	logger               *log.Logger
+	registryAuthFilePath string
+	imageStream          *imagev1.ImageStream
+}
+
+func NewExternalBinaryProvider(logger *log.Logger, releaseImage,
+	registryAuthfilePath string) (*ExternalBinaryProvider,
+	error) {
+	oc := util.NewCLIWithoutNamespace("default")
+
+	// Use a fixed cache or tmp directory for storing binaries
+	tmpDir := ""
+	binDir := pullSpecToDirName(releaseImage)
```

Comment:
Depending on how far you want to take this, it would be more efficient to create a cache directory based on the sha256 digest of the image the binary is coming from instead of considering the payload. This would allow you to safely share the binary extracted from payload A with payload B if the component image hasn't changed between them.

Reply:
I had thought of that but I wasn't sure if we always guarantee a release payload uses the image digest, instead of a tag pullspec. Is that true? I can always assume  it'll be `@sha256:<hash>`?

```
      {
        "name": "vsphere-csi-driver-syncer",
        "annotations": {
          "io.openshift.build.commit.id": "03b7e8ebf2c6347b40a340d892e1a66d47806300",
          "io.openshift.build.commit.ref": "",
          "io.openshift.build.source-location": "https://github.com/openshift/vmware-vsphere-csi-driver"
        },
        "from": {
          "kind": "DockerImage",
          "name": "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:d9c0515f4069cc18ce58fe200fb5eee73b703fba611c2b3892929e8f2754121d"
        },
        "generation": null,
        "importPolicy": {},
        "referencePolicy": {
          "type": ""
        }
      },
```